ethereum/contracts/BeefyClient.sol

```diff
@@ -0,0 +1,474 @@
+
+        // Verify that the leaf suppied by the relayer is part of the MMR
+        bytes32 leafHash = keccak256(encodeMMRLeaf(leaf));
+        require(
+            MMRProofVerification.verifyLeafProof(latestMMRRoot, leafHash, proof),
+            "Invalid leaf proof"
+        );
+
+        nextValidatorSet.id = leaf.nextAuthoritySetId;
+        nextValidatorSet.root = leaf.nextAuthoritySetRoot;
+        nextValidatorSet.length = leaf.nextAuthoritySetLen;
+    }
+
+    /* Private Functions */
+
+    function getValidatorSet(uint256 id) internal view returns (ValidatorSet memory) {
+        if (id == currentValidatorSet.id) {
+            return currentValidatorSet;
+        } else if (id == nextValidatorSet.id) {
+            return nextValidatorSet;
+        } else {
+            revert("unknown validator set");
+        }
+    }
+
+    /**
+     * @notice Deterministically generates a seed from the block hash at the block number of creation of the validation
+     * data plus MAXIMUM_NUM_SIGNERS
+     * @dev Note that `blockhash(blockNum)` will only work for the 256 most recent blocks. If
+     * `completeSignatureCommitment` is called too late, a new call to `newSignatureCommitment` is necessary to reset
+     * validation data's block number
+     * @param data a storage reference to the validationData struct
+     * @return onChainRandNums an array storing the random numbers generated inside this function
+     */
+    function getSeed(ValidationData storage data) private view returns (uint256) {
+        // @note Get payload.blocknumber, add BLOCK_WAIT_PERIOD
+        uint256 randomSeedBlockNum = data.blockNumber + BLOCK_WAIT_PERIOD;
+        // @note Create a hash seed from the block number
+        bytes32 randomSeedBlockHash = blockhash(randomSeedBlockNum);
+
+        return uint256(randomSeedBlockHash);
+    }
+
+    function requiredNumberOfSignatures(ValidatorSet memory vset) internal pure returns (uint256) {
+        return
+            (vset.length * THRESHOLD_NUMERATOR + THRESHOLD_DENOMINATOR - 1) / THRESHOLD_DENOMINATOR;
+    }
+
+    function verifyCommitment(
+        ValidationData storage data,
+        Commitment calldata commitment,
+        ValidatorProof calldata proof
+    ) internal view returns (bytes32) {
+        // Verify that sender is the same as in `newSignatureCommitment`
+        require(
+            msg.sender == data.senderAddress,
+            "Sender address does not match original validation data"
+        );
+
+        // Verify that block wait period has passed
+        require(block.number >= data.blockNumber + BLOCK_WAIT_PERIOD, "Block wait period not over");
+
+        // Check that payload.leaf.block_number is > last_known_block_number;
+        require(commitment.blockNumber > latestBeefyBlock, "Commitment blocknumber is too old");
+
+        ValidatorSet memory vset = getValidatorSet(commitment.validatorSetId);
```

Comment:
we could also have the BeefyClient inherit from a ValidatorSet contract instead of delegate to a ValidatorSet contract if the concern is related to multiple contract addresses/deployments rather than related to code style

Reply:
`getValidatorSet` no longer exists in the latest upload, so that simplifies things. Anyway I found it easier to reason about the validator set logic when its part of BeefyClient, rather than being scattered across multiple contracts.

Also simplifies deployment scripts.